business/mesh.go

```diff
@@ -249,33 +259,56 @@ func (in *MeshService) ResolveKialiControlPlaneCluster(r *http.Request) (*Cluste
 	return kialiControlPlaneCluster, nil
 }
 
+// convertKialiServiceToInstance converts a svc Service data structure of the
+// Kubernetes client to a KialiInstance data structure.
+func convertKialiServiceToInstance(svc *core_v1.Service) KialiInstance {
+	return KialiInstance{
+		ServiceName:      svc.Name,
+		Namespace:        svc.Namespace,
+		OperatorResource: svc.Annotations["operator-sdk/primary-resource"],
+		Version:          svc.Labels["app.kubernetes.io/version"],
+		Url:              svc.Annotations["kiali.io/external-url"],
+	}
+}
+
 // findKialiInNamespace tries to find a Kiali installation certain namespace of a cluster.
 // The clientSet argument should be an already initialized REST client to the API server of the
 // cluster. The namespace argument specifies the namespace where a Kiali instance will be looked for.
 // The clusterName argument is for logging purposes only.
-func findKialiInNamespace(namespace string, clusterName string, clientSet kubernetes.ClientInterface) (instances []KialiInstance) {
-	kialiNs, getNsErr := clientSet.GetNamespace(namespace)
+func findKialiInNamespace(namespace string, clusterName string, layer *Layer) (instances []KialiInstance) {
+	kialiNs, getNsErr := layer.Namespace.GetNamespace(namespace)
 	if getNsErr != nil && !errors.IsNotFound(getNsErr) {
 		log.Warningf("Discovery for Kiali instances in cluster [%s] failed: %s", clusterName, getNsErr.Error())
 		return
 	}
 	if kialiNs != nil {
 		// The operator and the helm charts set this fixed label. It's also
 		// present in the Istio addon manifest of Kiali.
-		services, getSvcErr := clientSet.GetServicesByLabels(kialiNs.Name, "app.kubernetes.io/name=kiali")
+		var services []core_v1.Service
+		var getSvcErr error
+		if IsNamespaceCached(kialiNs.Name) {
+			var tmpSvc []core_v1.Service
+			tmpSvc, getSvcErr = kialiCache.GetServices(kialiNs.Name, nil)
```

Comment:
Instead to add the filter logic here, you could reuse/add the filter logic into the kubernetes/filters.go

Reply:
OK, I moved the logic to kubernetes/filters.go
I hope this time all is OK.